File: case_studies/tennessee_star/pcalg_reference.R
#!/usr/bin/env Rscript

# Run the CRAN pcalg implementation of FCI+ on the committed STAR panels.
#
# The application deliberately supplies a custom G-square test that mirrors
# src/fci_engine/ci/discrete.py. In particular, it does not use pcalg's
# disCItest n.min shortcut, which can declare high-dimensional tables
# independent before calculating the statistic. This keeps the statistical
# decision rule aligned across the Python and R implementations.

suppressPackageStartupMessages(library(pcalg))

parse_args <- function(arguments) {
  parsed <- list(
    input_dir = "case_studies/tennessee_star/data/processed",
    output_dir = "case_studies/tennessee_star/output",
    alpha = 0.05,
    benchmark_repeats = 3L,
    order_audit = TRUE
  )
  index <- 1L
  while (index <= length(arguments)) {
    flag <- arguments[[index]]
    if (flag == "--no-order-audit") {
      parsed$order_audit <- FALSE
      index <- index + 1L
      next
    }
    if (index == length(arguments)) {
      stop(paste("Missing value for", flag))
    }
    value <- arguments[[index + 1L]]
    if (flag == "--input-dir") {
      parsed$input_dir <- value
    } else if (flag == "--output-dir") {
      parsed$output_dir <- value
    } else if (flag == "--alpha") {
      parsed$alpha <- as.numeric(value)
    } else if (flag == "--benchmark-repeats") {
      parsed$benchmark_repeats <- as.integer(value)
    } else {
      stop(paste("Unknown argument:", flag))
    }
    index <- index + 2L
  }
  if (
    is.na(parsed$alpha) ||
    parsed$alpha <= 0 ||
    parsed$alpha >= 1
  ) {
    stop("--alpha must be between zero and one")
  }
  if (
    is.na(parsed$benchmark_repeats) ||
    parsed$benchmark_repeats < 1L
  ) {
    stop("--benchmark-repeats must be positive")
  }
  parsed
}

compact_g_square <- function(table) {
  row_keep <- rowSums(table) > 0
  col_keep <- colSums(table) > 0
  compact <- table[row_keep, col_keep, drop = FALSE]
  if (nrow(compact) < 2L || ncol(compact) < 2L) {
    return(c(statistic = 0, degrees_freedom = 0))
  }

  total <- sum(compact)
  expected <- outer(rowSums(compact), colSums(compact)) / total
  valid <- compact > 0 & expected > 0
  statistic <- 2 * sum(compact[valid] * log(compact[valid] / expected[valid]))
  degrees_freedom <- (nrow(compact) - 1L) * (ncol(compact) - 1L)
  c(statistic = statistic, degrees_freedom = degrees_freedom)
}

python_equivalent_g_square <- function(x, y, S, suffStat) {
  data <- suffStat$dm
  x_levels <- 0:(suffStat$nlev[[x]] - 1L)
  y_levels <- 0:(suffStat$nlev[[y]] - 1L)

  if (length(S) == 0L) {
    strata <- list(seq_len(nrow(data)))
  } else {
    conditioning <- data[, S, drop = FALSE]
    keys <- apply(conditioning, 1L, paste, collapse = "\r")
    strata <- split(seq_len(nrow(data)), keys, drop = TRUE)
  }

  statistic <- 0
  degrees_freedom <- 0L
  for (rows in strata) {
    contingency <- table(
      factor(data[rows, x], levels = x_levels),
      factor(data[rows, y], levels = y_levels)
    )
    part <- compact_g_square(contingency)
    statistic <- statistic + part[["statistic"]]
    degrees_freedom <- degrees_freedom + part[["degrees_freedom"]]
  }

  if (degrees_freedom <= 0L) {
    return(1)
  }
  pchisq(statistic, df = degrees_freedom, lower.tail = FALSE)
}

fit_pcalg_fci_plus <- function(data, alpha) {
  counter <- 0L
  counted_test <- function(x, y, S, suffStat) {
    counter <<- counter + 1L
    python_equivalent_g_square(x, y, S, suffStat)
  }
  sufficient_statistics <- list(
    dm = data,
    nlev = apply(data, 2L, max) + 1L
  )
  warnings <- character()
  elapsed <- system.time(
    fit <- withCallingHandlers(
      fciPlus(
        suffStat = sufficient_statistics,
        indepTest = counted_test,
        alpha = alpha,
        labels = colnames(data),
        verbose = FALSE,
        selectionBias = TRUE
      ),
      warning = function(condition) {
        warnings <<- c(warnings, conditionMessage(condition))
        invokeRestart("muffleWarning")
      }
    )
  )[["elapsed"]]
  list(
    fit = fit,
    ci_tests = counter,
    elapsed_seconds = as.numeric(elapsed),
    warnings = unique(warnings)
  )
}

endpoint_name <- function(mark) {
  marks <- c("1" = "CIRCLE", "2" = "ARROW", "3" = "TAIL")
  unname(marks[[as.character(mark)]])
}

edge_notation <- function(x, y, endpoint_x, endpoint_y) {
  symbols <- c(
    "CIRCLE:CIRCLE" = "o-o",
    "CIRCLE:ARROW" = "o->",
    "ARROW:CIRCLE" = "<-o",
    "TAIL:ARROW" = "-->",
    "ARROW:TAIL" = "<--",
    "ARROW:ARROW" = "<->",
    "TAIL:TAIL" = "---",
    "CIRCLE:TAIL" = "o--",
    "TAIL:CIRCLE" = "--o"
  )
  key <- paste(endpoint_x, endpoint_y, sep = ":")
  paste(x, symbols[[key]], y)
}

edge_frame <- function(fit, canonical_labels, panel) {
  matrix <- fit@amat
  labels <- colnames(matrix)
  order_index <- setNames(seq_along(canonical_labels), canonical_labels)
  rows <- list()
  row_index <- 1L
  if (ncol(matrix) > 1L) {
    for (i in seq_len(ncol(matrix) - 1L)) {
      for (j in seq.int(i + 1L, ncol(matrix))) {
        if (matrix[i, j] == 0L && matrix[j, i] == 0L) {
          next
        }
        x <- labels[[i]]
        y <- labels[[j]]
        endpoint_x <- endpoint_name(matrix[j, i])
        endpoint_y <- endpoint_name(matrix[i, j])
        if (order_index[[x]] > order_index[[y]]) {
          temporary_label <- x
          x <- y
          y <- temporary_label
          temporary_endpoint <- endpoint_x
          endpoint_x <- endpoint_y
          endpoint_y <- temporary_endpoint
        }
        rows[[row_index]] <- data.frame(
          panel = panel,
          algorithm = "pcalg_fci_plus",
          x = x,
          y = y,
          endpoint_x = endpoint_x,
          endpoint_y = endpoint_y,
          edge = edge_notation(x, y, endpoint_x, endpoint_y),
          stringsAsFactors = FALSE
        )
        row_index <- row_index + 1L
      }
    }
  }
  if (length(rows) == 0L) {
    return(data.frame(
      panel = character(),
      algorithm = character(),
      x = character(),
      y = character(),
      endpoint_x = character(),
      endpoint_y = character(),
      edge = character(),
      stringsAsFactors = FALSE
    ))
  }
  do.call(rbind, rows)
}

edge_signature <- function(edges) {
  if (nrow(edges) == 0L) {
    return(character())
  }
  sort(paste(
    edges$x,
    edges$y,
    edges$endpoint_x,
    edges$endpoint_y,
    sep = "|"
  ))
}

skeleton_signature <- function(edges) {
  if (nrow(edges) == 0L) {
    return(character())
  }
  sort(paste(edges$x, edges$y, sep = "|"))
}

jaccard <- function(left, right) {
  union_values <- union(left, right)
  if (length(union_values) == 0L) {
    return(1)
  }
  length(intersect(left, right)) / length(union_values)
}

run_panel <- function(panel, path, alpha, benchmark_repeats, order_audit) {
  frame <- read.csv(path, check.names = FALSE)
  data <- as.matrix(frame)
  storage.mode(data) <- "integer"
  canonical_labels <- colnames(data)

  fits <- lapply(
    seq_len(benchmark_repeats),
    function(unused) fit_pcalg_fci_plus(data, alpha)
  )
  primary <- fits[[1L]]
  primary_edges <- edge_frame(primary$fit, canonical_labels, panel)
  all_warnings <- unique(unlist(lapply(fits, function(item) item$warnings)))

  audit_rows <- list()
  if (order_audit) {
    primary_edge_signature <- edge_signature(primary_edges)
    primary_skeleton <- skeleton_signature(primary_edges)
    for (shift in seq_along(canonical_labels) - 1L) {
      if (shift == 0L) {
        audit_fit <- primary
        audit_edges <- primary_edges
      } else {
        permutation <- c(
          seq.int(shift + 1L, length(canonical_labels)),
          seq_len(shift)
        )
        audit_data <- data[, permutation, drop = FALSE]
        audit_fit <- fit_pcalg_fci_plus(audit_data, alpha)
        audit_edges <- edge_frame(audit_fit$fit, canonical_labels, panel)
        all_warnings <- unique(c(all_warnings, audit_fit$warnings))
      }
      audit_rows[[shift + 1L]] <- data.frame(
        panel = panel,
        ordering = paste0("cyclic_shift_", shift),
        first_variable = colnames(
          data[, c(
            seq.int(shift + 1L, length(canonical_labels)),
            if (shift > 0L) seq_len(shift) else integer()
          ), drop = FALSE]
        )[[1L]],
        exact_pag_match = identical(
          edge_signature(audit_edges),
          primary_edge_signature
        ),
        skeleton_jaccard = jaccard(
          skeleton_signature(audit_edges),
          primary_skeleton
        ),
        edges = nrow(audit_edges),
        ci_tests = audit_fit$ci_tests,
        elapsed_seconds = audit_fit$elapsed_seconds,
        stringsAsFactors = FALSE
      )
    }
  } else {
    audit_rows[[1L]] <- data.frame(
      panel = panel,
      ordering = "not_run",
      first_variable = canonical_labels[[1L]],
      exact_pag_match = NA,
      skeleton_jaccard = NA_real_,
      edges = nrow(primary_edges),
      ci_tests = primary$ci_tests,
      elapsed_seconds = primary$elapsed_seconds,
      stringsAsFactors = FALSE
    )
  }
  audit <- do.call(rbind, audit_rows)

  run <- data.frame(
    panel = panel,
    algorithm = "pcalg_fci_plus",
    samples = nrow(data),
    nodes = ncol(data),
    edges = nrow(primary_edges),
    ci_tests = primary$ci_tests,
    median_elapsed_seconds = median(vapply(
      fits,
      function(item) item$elapsed_seconds,
      numeric(1)
    )),
    elapsed_runs_seconds = paste(vapply(
      fits,
      function(item) item$elapsed_seconds,
      numeric(1)
    ), collapse = ";"),
    r_version = as.character(getRversion()),
    pcalg_version = as.character(packageVersion("pcalg")),
    ci_test = "Python-equivalent G-square",
    alpha = alpha,
    selection_bias = TRUE,
    orderings_checked = nrow(audit),
    exact_pag_match_rate = mean(audit$exact_pag_match, na.rm = TRUE),
    mean_skeleton_jaccard = mean(audit$skeleton_jaccard, na.rm = TRUE),
    warning_count = length(all_warnings),
    warnings = paste(all_warnings, collapse = " | "),
    stringsAsFactors = FALSE
  )
  list(run = run, edges = primary_edges, audit = audit)
}

arguments <- parse_args(commandArgs(trailingOnly = TRUE))
dir.create(arguments$output_dir, recursive = TRUE, showWarnings = FALSE)

panel_files <- c(
  attrition = "attrition_panel.csv",
  longitudinal = "longitudinal_panel.csv",
  focused_treatment = "focused_treatment_panel.csv"
)
results <- lapply(names(panel_files), function(panel) {
  run_panel(
    panel = panel,
    path = file.path(arguments$input_dir, panel_files[[panel]]),
    alpha = arguments$alpha,
    benchmark_repeats = arguments$benchmark_repeats,
    order_audit = arguments$order_audit
  )
})

runs <- do.call(rbind, lapply(results, function(item) item$run))
edges <- do.call(rbind, lapply(results, function(item) item$edges))
audit <- do.call(rbind, lapply(results, function(item) item$audit))

write.csv(
  runs,
  file.path(arguments$output_dir, "star_pcalg_runs.csv"),
  row.names = FALSE
)
write.csv(
  edges,
  file.path(arguments$output_dir, "star_pcalg_edges.csv"),
  row.names = FALSE
)
write.csv(
  audit,
  file.path(arguments$output_dir, "star_pcalg_order_audit.csv"),
  row.names = FALSE
)

cat("R version:", as.character(getRversion()), "\n")
cat("pcalg version:", as.character(packageVersion("pcalg")), "\n")
cat("runs:", normalizePath(
  file.path(arguments$output_dir, "star_pcalg_runs.csv")
), "\n")
cat("edges:", normalizePath(
  file.path(arguments$output_dir, "star_pcalg_edges.csv")
), "\n")
cat("order audit:", normalizePath(
  file.path(arguments$output_dir, "star_pcalg_order_audit.csv")
), "\n")

File: case_studies/tennessee_star/pcalg_reference.py
"""Load and optionally refresh the independent CRAN pcalg FCI+ benchmark."""

from __future__ import annotations

import shutil
import subprocess
from pathlib import Path
from typing import Any

import pandas as pd

from case_studies.tennessee_star.study import (
    OUTPUT_DIR,
    PROCESSED_DIR,
    TEMPORAL_TIERS,
    PreparedStudy,
)

HERE = Path(__file__).resolve().parent
R_SCRIPT = HERE / "pcalg_reference.R"
PCALG_RUNS_PATH = OUTPUT_DIR / "star_pcalg_runs.csv"
PCALG_EDGES_PATH = OUTPUT_DIR / "star_pcalg_edges.csv"
PCALG_ORDER_AUDIT_PATH = OUTPUT_DIR / "star_pcalg_order_audit.csv"
PCALG_ALGORITHM = "pcalg_fci_plus"


def find_rscript(explicit: Path | None = None) -> Path | None:
    """Return an available Rscript executable."""

    if explicit is not None:
        candidate = explicit.expanduser().resolve()
        if not candidate.exists():
            raise FileNotFoundError(f"Rscript does not exist: {candidate}")
        return candidate

    discovered = shutil.which("Rscript")
    if discovered is not None:
        return Path(discovered)
    for candidate in (
        Path("/opt/homebrew/bin/Rscript"),
        Path("/usr/local/bin/Rscript"),
        Path("/Library/Frameworks/R.framework/Resources/bin/Rscript"),
    ):
        if candidate.exists():
            return candidate
    return None


def refresh_pcalg_reference(
    *,
    output_directory: Path = OUTPUT_DIR,
    rscript: Path | None = None,
    alpha: float = 0.05,
    benchmark_repeats: int = 3,
    order_audit: bool = True,
    timeout: int = 900,
) -> None:
    """Run the committed R benchmark and replace its CSV artifacts."""

    executable = find_rscript(rscript)
    if executable is None:
        raise RuntimeError(
            "Rscript is not available. Install R and CRAN pcalg, or pass "
            "--rscript explicitly."
        )
    command = [
        str(executable),
        str(R_SCRIPT),
        "--input-dir",
        str(PROCESSED_DIR),
        "--output-dir",
        str(output_directory),
        "--alpha",
        str(alpha),
        "--benchmark-repeats",
        str(benchmark_repeats),
    ]
    if not order_audit:
        command.append("--no-order-audit")
    completed = subprocess.run(
        command,
        capture_output=True,
        text=True,
        timeout=timeout,
        check=False,
    )
    if completed.returncode != 0:
        details = completed.stderr.strip() or completed.stdout.strip()
        raise RuntimeError(f"pcalg FCI+ benchmark failed: {details[-2000:]}")


def load_pcalg_reference(
    study: PreparedStudy,
    *,
    output_directory: Path = OUTPUT_DIR,
    expected_alpha: float = 0.05,
) -> tuple[list[dict[str, Any]], dict[str, Any]]:
    """Load validated pcalg run records for inclusion in the report payload."""

    runs_path = output_directory / PCALG_RUNS_PATH.name
    edges_path = output_directory / PCALG_EDGES_PATH.name
    audit_path = output_directory / PCALG_ORDER_AUDIT_PATH.name
    missing = [
        path for path in (runs_path, edges_path, audit_path) if not path.exists()
    ]
    if missing:
        names = ", ".join(path.name for path in missing)
        raise FileNotFoundError(
            f"Missing pcalg benchmark artifacts: {names}. Run "
            "`Rscript case_studies/tennessee_star/pcalg_reference.R`."
        )

    run_frame = pd.read_csv(runs_path)
    edge_frame = pd.read_csv(edges_path)
    audit_frame = pd.read_csv(audit_path)
    records: list[dict[str, Any]] = []
    for row in run_frame.to_dict(orient="records"):
        panel_name = str(row["panel"])
        if panel_name not in study.panels:
            raise ValueError(f"Unknown pcalg panel: {panel_name}")
        panel = study.panels[panel_name]
        if int(row["samples"]) != len(panel.data):
            raise ValueError(f"Stale pcalg sample count for {panel_name}.")
        if int(row["nodes"]) != len(panel.data.columns):
            raise ValueError(f"Stale pcalg node count for {panel_name}.")
        if abs(float(row["alpha"]) - expected_alpha) > 1e-12:
            raise ValueError(f"Stale pcalg alpha for {panel_name}.")

        panel_edges = edge_frame.loc[edge_frame["panel"] == panel_name, :]
        edge_records = [
            {
                "x": str(edge["x"]),
                "y": str(edge["y"]),
                "endpoint_x": str(edge["endpoint_x"]),
                "endpoint_y": str(edge["endpoint_y"]),
                "edge": str(edge["edge"]),
                "bootstrap_frequency": None,
            }
            for edge in panel_edges.to_dict(orient="records")
        ]
        temporal_flags = _temporal_flags(edge_records)
        panel_audit = audit_frame.loc[audit_frame["panel"] == panel_name, :]
        elapsed_runs = [
            float(value)
            for value in str(row["elapsed_runs_seconds"]).split(";")
            if value
        ]
        warnings = _optional_text(row.get("warnings"))
        records.append(
            {
                "panel": panel_name,
                "algorithm": PCALG_ALGORITHM,
                "samples": int(row["samples"]),
                "nodes": int(row["nodes"]),
                "node_names": list(panel.data.columns),
                "edges": int(row["edges"]),
                "ci_tests": int(row["ci_tests"]),
                "median_elapsed_seconds": float(row["median_elapsed_seconds"]),
                "elapsed_runs_seconds": elapsed_runs,
                "temporal_flags": temporal_flags,
                "pag_edges": edge_records,
                "bootstrap_adjacencies": [],
                "assumption_notes": [
                    "Independent reference implementation from CRAN pcalg; "
                    "external package output is not treated as ground truth.",
                    "The R runner supplies the same compact-table G-square "
                    "decision rule as the Python implementation.",
                    "pcalg::fciPlus does not expose the local implementation's "
                    "explicit k sparsity-bound parameter.",
                ],
                "implementation": {
                    "language": "R",
                    "r_version": str(row["r_version"]),
                    "package": "pcalg",
                    "package_version": str(row["pcalg_version"]),
                    "ci_test": str(row["ci_test"]),
                    "selection_bias": bool(row["selection_bias"]),
                    "warning_count": int(row["warning_count"]),
                    "warnings": warnings,
                },
                "order_audit": {
                    "orderings_checked": int(row["orderings_checked"]),
                    "exact_pag_match_rate": float(row["exact_pag_match_rate"]),
                    "mean_skeleton_jaccard": float(row["mean_skeleton_jaccard"]),
                    "rows": [
                        {
                            "ordering": str(item["ordering"]),
                            "first_variable": str(item["first_variable"]),
                            "exact_pag_match": bool(item["exact_pag_match"]),
                            "skeleton_jaccard": float(item["skeleton_jaccard"]),
                            "edges": int(item["edges"]),
                            "ci_tests": int(item["ci_tests"]),
                            "elapsed_seconds": float(item["elapsed_seconds"]),
                        }
                        for item in panel_audit.to_dict(orient="records")
                    ],
                },
            }
        )

    metadata = {
        "algorithm": PCALG_ALGORITHM,
        "r_script": _portable_path(R_SCRIPT),
        "runs_csv": _portable_path(runs_path),
        "edges_csv": _portable_path(edges_path),
        "order_audit_csv": _portable_path(audit_path),
        "r_version": str(run_frame.iloc[0]["r_version"]),
        "pcalg_version": str(run_frame.iloc[0]["pcalg_version"]),
        "ci_test": str(run_frame.iloc[0]["ci_test"]),
        "alpha": float(run_frame.iloc[0]["alpha"]),
    }
    return records, metadata


def _temporal_flags(edges: list[dict[str, Any]]) -> list[str]:
    flags = []
    for edge in edges:
        x = str(edge["x"])
        y = str(edge["y"])
        endpoint_x = str(edge["endpoint_x"])
        endpoint_y = str(edge["endpoint_y"])
        if endpoint_x == "TAIL" and endpoint_y == "ARROW":
            source, target = x, y
        elif endpoint_x == "ARROW" and endpoint_y == "TAIL":
            source, target = y, x
        else:
            continue
        if TEMPORAL_TIERS.get(source, 0) > TEMPORAL_TIERS.get(target, 0):
            flags.append(f"{source} --> {target}")
    return sorted(flags)


def _optional_text(value: object) -> str:
    if value is None or pd.isna(value):
        return ""
    return str(value)


def _portable_path(path: Path) -> str:
    repository_root = HERE.parents[1]
    try:
        return str(path.resolve().relative_to(repository_root))
    except ValueError:
        return str(path.resolve())
